### PR TITLE
Implement `java.net` for Windows

### PR DESCRIPTION
**Pull request by WojciechMazur**
This PR continues series of Windows support changes and it adds proper support for `java.net` package. 

* Fixed implementation of java.net to run existing tests on Windows
* Moved `SocketHelpers` to `javalib` from `posixlib` and adapted sripted tests using this package-private file
* Added missing bindings needed to support `java.net`
* Provided platform depended implementation of `so_linger` struct. 
* Adjusted unit test for platform depended code.

### CODE REVIEW

File: javalib/src/main/scala/java/net/AbstractPlainSocketImpl.scala [hunk 1]
```diff
@@ -454,25 +479,34 @@ private[net] abstract class AbstractPlainSocketImpl extends SocketImpl {
...
         val ptr = stackalloc[CInt]
         !ptr = value.asInstanceOf[Int]
         ptr.asInstanceOf[Ptr[Byte]]
     }
 
-    if (socket.setsockopt(fd.fd, level, optValue, opt, len) == -1) {
+    if (socket.setsockopt(fd.fd, level, optValue, opt, len) != 0) {
+      val errCode =
+        if (isWindows) WSAGetLastError()
+        else errno.errno
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/AbstractPlainSocketImpl.scala:506]: This fragment is repeated a few times. Consider putting it in a reusable private def.

File: javalib/src/main/scala/java/net/SocketHelpers.scala [hunk 2]
```diff
@@ -69,20 +80,23 @@ object SocketHelpers {
         time.tv_sec = tv_sec
         time.tv_usec = tv_usec
 
-        if (connect(sock, ai.ai_addr, ai.ai_addrlen) != 0) {
-          return false
+        connect(sock, ai.ai_addr, ai.ai_addrlen) match {
+          case 0 => ()
+          case error =>
+            return false
         }
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/SocketHelpers.scala:87]: This change doesn't seem warranted.

File: javalib/src/main/scala/java/net/SocketHelpers.scala [hunk 3]
```diff
@@ -95,20 +109,20 @@ object SocketHelpers {
...
         time.tv_sec = tv_sec
         time.tv_usec = tv_usec
-
-        if (select(sock + 1, fdsetPtr, null, null, time) != 1) {
-          return false
-        } else {
-          val buf = stackalloc[CChar](5.toUInt)
-          val recBytes = recv(sock, buf, 5.toUInt, 0)
-          if (recBytes < 4) {
+        select(sock + 1, fdsetPtr, null, null, time) match {
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/SocketHelpers.scala:112]: Same here.

File: javalib/src/main/scala/java/net/WindowsPlainSocketImpl.scala [hunk 4]
```diff
@@ -109,10 +109,13 @@ private[net] class WindowsPlainSocketImpl extends AbstractPlainSocketImpl {
   ): Unit = {
     val mode = stackalloc[Int]
     if (blocking)
-      !mode = 1
-    else
       !mode = 0
-    ioctlSocket(fd.handle, FIONBIO, mode)
+    else
+      !mode = 1
```
Comment (sjrd) [inline javalib/src/main/scala/java/net/WindowsPlainSocketImpl.scala:114]: Why did the value of `mode` get inverted?

**Review by sjrd — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/net/WindowsPlainSocketImpl.scala (hunk 4 above)
Comment (WojciechMazur) [reply javalib/src/main/scala/java/net/WindowsPlainSocketImpl.scala:114]: Previous usage of mode was incorrect, I have misinterpreted the documentation of ioctlSocket. 

Fragment from Microsoft [documentation of ioctlSocket](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-ioctlsocket):
```
//-------------------------
// Set the socket I/O mode: In this case FIONBIO
// enables or disables the blocking mode for the 
// socket based on the numerical value of iMode.
// If iMode = 0, blocking is enabled; 
// If iMode != 0, non-blocking mode is enabled.

iResult = ioctlsocket(m_socket, FIONBIO, &iMode);

```

File: javalib/src/main/scala/java/net/URI.scala [hunk 5]
```diff
@@ -91,9 +106,7 @@ final class URI private () extends Comparable[URI] with Serializable {
       earlyStop = true
     }
     if (!earlyStop) {
-      if (scheme != null && path != null &&
-          path.length() > 0 &&
-          path.charAt(0) != '/') {
+      if (scheme != null && isRelativePath(path)) {
         throw new URISyntaxException(path, "Relative path")
       }
```
[the remark this answers is not in the document]
Comment (WojciechMazur) [reply javalib/src/main/scala/java/net/URI.scala]: I've tried to revert changes in the `URI`, it looks like it's passing all the current tests. Probably these changes are no longer needed, as the underlying issues were probably already fixed in `java.io.File` or `java.nio.file.Path` implementations. 
Thank you for spotting this!

**Review by sjrd — APPROVED** [review]

**Pull request merged**